src/SAWScript/Yosys/CompositionalTranslation.hs

```diff
@@ -0,0 +1,343 @@
+
+import qualified Cryptol.TypeCheck.Type as C
+
+import SAWScript.Panic (panic)
+
+import SAWScript.Yosys.Utils
+import SAWScript.Yosys.IR
+import SAWScript.Yosys.Cell
+
+type ModuleName = Text
+type CellName = Text
+type Pattern = [Bitrep]
+type PatternMap m = Map Pattern ((YosysBitvecConsumer -> Pattern -> m SC.Term) -> m SC.Term)
+
+data CellStateInfo = CellStateInfo
+  { _cellStateInfoType :: SC.Term -- cell state type - either a bitvector for a $dff, or a record type
+  , _cellStateInfoCryptolType :: C.Type -- cryptol type for the above
+  , _cellStateInfoFields :: Maybe (Map Text (SC.Term, C.Type)) -- if the type is a record, the fields of the record
+  }
+makeLenses ''CellStateInfo
+
+data TranslatedModule = TranslatedModule
+  { _translatedModuleStateInfo :: Maybe CellStateInfo -- information about the state type for this module
+  , _translatedModuleTerm :: SC.Term -- the lambda term for the output record (including state) in terms of the inputs (including state)
+  , _translatedModuleType :: SC.Term
+  , _translatedModuleCryptolType :: C.Type
+  }
+makeLenses ''TranslatedModule
+
+data TranslationContext m = TranslationContext
+  { _translationContextModules :: Map ModuleName TranslatedModule
+  , _translationContextStateTypes :: Map CellName CellStateInfo -- state type for every stateful cell in this module (including sequential submodules)
+  , _translationContextPatternMap :: PatternMap m -- for each pattern, a term representing that pattern (parameterized by a function to get a term representing any other pattern)
+  }
+makeLenses ''TranslationContext
+
+buildTranslationContextStateTypes ::
+  MonadIO m =>
+  SC.SharedContext ->
+  Map ModuleName TranslatedModule ->
+  Module ->
+  m (Map CellName CellStateInfo)
+buildTranslationContextStateTypes sc mods m = do
+  fmap (Map.mapMaybe id) . forM (m ^. moduleCells) $ \c -> do
+    case Map.lookup (c ^. cellType) mods of
+      Just tm -> pure $ tm ^. translatedModuleStateInfo
+      Nothing -> case c ^. cellType of
+        "$dff" | Just w <- length <$> Map.lookup "Q" (c ^. cellConnections) -> do
+          _cellStateInfoType <- liftIO . SC.scBitvector sc $ fromIntegral w
+          let _cellStateInfoCryptolType = C.tWord $ C.tNum w
+          let _cellStateInfoFields = Nothing
+          pure $ Just CellStateInfo{..}
+        _ -> pure Nothing
+
+lookupStateFor ::
+  forall m.
+  MonadIO m =>
+  SC.SharedContext ->
+  Map CellName CellStateInfo -> -- state type info for each cell
+  SC.Term -> -- record term mapping (zenc-ed) cell names to cell states
+  CellName -> -- cell state to lookup
+  m SC.Term
+lookupStateFor sc states inpst cnm = do
+  let fieldnm = cellIdentifier cnm
+  cryptolRecordSelect sc (Map.mapKeys cellIdentifier states) inpst fieldnm
+
+-- | Add a record-typed field named __states__ to the given mapping of field names to types.
+insertStateField ::
+  MonadIO m =>
+  SC.SharedContext ->
+  Map Text (SC.Term, C.Type) {- ^ The field types of "__states__" -} ->
```

Comment:
Haddock formatting nit pick:

```suggestion
  Map Text (SC.Term, C.Type) {- ^ The field types of \"__states__\" -} ->
```

Without escaping the double quotes, Haddock would try to render `__states__` as a Haskell module name.